Mcc 222759 add self registration

ssucharitdas opened the pull request:
PR includes:
- 4 View Controllers (Email, Password, Security Question and Security Answer) to create an account without following Rave route.
- Logic to display both Rave and S3 based forms
- Make a registration API call to Babbage
- Documentation updates

Please review and provide feedback.

ssucharitdas committed:
Using a specific version of the pod rather than going to latest available.

ssucharitdas committed:
Adding Email Page layout.

ssucharitdas committed:
Adding the desired pages for Registration workflow.

ssucharitdas committed:
Fixing the row selection from table view and adding the pbx file

ssucharitdas committed:
Updating the EmailViewController with needful logic.

ssucharitdas committed:
Password field confirmation and logic.

ssucharitdas committed:
Passing values across segues (screens) and registration call to Babbage for new user.

ssucharitdas committed:
Correcting the logic on CreateAccount to send segue or print the error, and minor typo corrected

ssucharitdas committed:
Updating the pod to same as develop to make it easier for testing

ssucharitdas committed:
Updating code documentation and Readme.

ssucharitdas committed:
Code cleanup.

ssucharitdas committed:
Update the tokens and correcting the logic.

ssucharitdas commented:
@ncarroll-mdsol Please review, I will keep updating as we move along. [I have no DNM tags here so added in the PR itself]

ssucharitdas committed:
Updating the segue to run on main thread and not on another thread for Creating Account

ssucharitdas committed:
Updating the nav title, the data saving mechanism.

ssucharitdas committed:
Using client and userid to collect the data.

ssucharitdas committed:
Update the segue to be performed after value setting.

ssucharitdas committed:
Updating the Image saving api calls and making the save button operational only when subjects are loaded for the user.

ssucharitdas committed:
Switching on/off save button based on the loading of subjects.

ncarroll-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -11,26 +11,68 @@ import UIKit
+                // Check all subjects loaded
+                if error != nil {
+                    loadedSubjectsAndErrors.append(error)
+                    if subjects != nil {
```
Will there ever be a case where both `error` and `subjects` are not nil? What does the logic within the `error != nil` conditional do?

ncarroll-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -11,26 +11,68 @@ import UIKit
+                else {
+                    // Enable image saving button once loaded
+                    self.collectedSubjects = subjects as! [MDSubject]
+                    dispatch_async(dispatch_get_main_queue(), { 
```
We should be consistent about how we handle threading - I'd change this to use the NSOperationQueue.

ncarroll-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -48,12 +90,17 @@ class CaptureImageViewController: UIViewController, UIImagePickerControllerDeleg
+                    NSOperationQueue.mainQueue().addOperationWithBlock {
+                        if err == nil {
+                            print(err == nil ? "Data Saved" : err?.description)
+                        }
```
I think this ternary should be pulled out of the conditional? Or else one of the conditions will never be reached.

ssucharitdas replied:
Converting to an alert.

ncarroll-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -48,12 +90,17 @@ class CaptureImageViewController: UIViewController, UIImagePickerControllerDeleg
-                print(err == nil ? "Data Saved" : err?.description);
-                self.imageView.image = nil
-            })
+            for subject in self.collectedSubjects {
```
I think for the sample app purposes we'll only want to collectData for a single subject. In most cases there will only be one, but it might make more sense to just store one when the view loads, and then use it here (like you're doing with collectedSubjects)

ncarroll-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -11,26 +11,68 @@ import UIKit
+        bgQueue.addOperationWithBlock() {
+            let clientFactory = MDClientFactory.sharedInstance()
+            let client = clientFactory.clientOfType(MDClientType.Network);
+            let user = self.datastore.userWithID(Int64(self.userID))
```
There may be some issues with using a datastore created on a different thread. I think elsewhere in the app we create a new datastore within the background block.

ncarroll-mdsol replied:
Also, is the Int64 conversion necessary? It looks like the var is defined already as an Int64

ssucharitdas committed:
Keeping a check on version of IQKeyboard.

ssucharitdas committed:
Version addition to the Keyboard manager.

ssucharitdas committed:
Update based on PR comments.

ssucharitdas committed:
Comment based updates.

ssucharitdas committed:
Update babbage to use new self registration api.

ssucharitdas committed:
Make the previous screen selected security appear on Create Account screen.

ssucharitdas committed:
Updating the functions to use send data.

ncarroll-mdsol committed:
MCC-222759 - update pods, clena up code

ncarroll-mdsol committed:
MCC-222759 - move vars back in bgqueue

kbohlmann-mdsol commented on AppConnectSwift/SecurityQuestionViewController.swift:
```diff
@@ -0,0 +1,66 @@
+        }
+    }
+    
+}
```
nitpick : EOL @ EOF

kbohlmann-mdsol commented on AppConnectSwift/AppDelegate.swift:
```diff
@@ -17,7 +17,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         let dir = NSFileManager.defaultManager().URLsForDirectory(NSSearchPathDirectory.DocumentDirectory, inDomains: NSSearchPathDomainMask.UserDomainMask).last
         let key = "12345678901234567890123456789012".dataUsingEncoding(NSUTF8StringEncoding)
         // TODO: Substitute the apiToken value with proper token
```
This TODO is done, looks like?

ncarroll-mdsol replied:
We'll actually want to remove the token here, and invalidate it. This shouldn't have been committed in the first place.

ssucharitdas replied:
Removed the token!

kbohlmann-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -11,14 +11,22 @@ import UIKit
 class CaptureImageViewController: UIViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
     
     @IBOutlet weak var imageView: UIImageView!
+    @IBOutlet weak var saveImage: UIBarButtonItem!
```
These two variable names are not great. saveImage more than imageView -- is that a button? 
maybe: "saveImageButton" ?

kbohlmann-mdsol commented on AppConnectSwift/AppDelegate.swift:
```diff
@@ -17,7 +17,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         let key = "12345678901234567890123456789012".dataUsingEncoding(NSUTF8StringEncoding)
         // TODO: Substitute the apiToken value with proper token
-        MDBabbage.startWithEnvironment(MDClientEnvironment.Validation, apiToken: "Random String", publicDirectory: dir, privateDirectory: dir, encryptionKey: key)
+        MDBabbage.startWithEnvironment(MDClientEnvironment.Sandbox, apiToken: "3d4a7f259255a9b9c69bad748447ee64", publicDirectory: dir, privateDirectory: dir, encryptionKey: key)
```
Wait, we shouldn't have an actual token in the repo. This is a public facing repo.
This is for the final user to put in token, and the TODO should stay as a reminder to them.

kbohlmann-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -31,6 +39,7 @@ class CaptureImageViewController: UIViewController, UIImagePickerControllerDeleg
         self.dismissViewControllerAnimated(true, completion: nil)
         self.image = info[UIImagePickerControllerOriginalImage] as! UIImage
         imageView.image = self.image
+        self.data = self.compressFile() as? NSData
```
Why are we compressing this? We probably want the image uploaded to be the same as the image the user sees on the screen... and, it doesn't look this compresses (or operates on a file), it actually just reduces image quality, if we need it could rename to "scaleDownImage".

ssucharitdas replied:
@kbohlmann-mdsol 
We have a cap of < 10Mb and we might cross that limit with some images. It actually scales down keeping the aspect ratio proper for the image to be saved.
Let me know we want it to be name to scaleDownImage.

kbohlmann-mdsol replied:
ok. So the method actually takes an image and scales it down, and converts it to an nsdata. 
Try to come up with a name that lets us know what is going to happen? 
Like, scaleDownAndConvertImage or something along those lines

kbohlmann-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -48,12 +57,32 @@ class CaptureImageViewController: UIViewController, UIImagePickerControllerDeleg
+                                    NSOperationQueue.mainQueue().addOperationWithBlock {
+                                        self.showAlert("Save Image", message: "Data saved successfully")
+                                    }
+                                    datastore = nil
```
are all these explicit = nil's necessary?

ssucharitdas committed:
Update w.r.t comments.

ssucharitdas committed:
Updating the method signature of collectData.

ssucharitdas committed:
Resolving conflicts

ssucharitdas committed:
Renaming method to make it go with operation it is handling.

ssucharitdas committed:
Bumping pod spec to 80.

ssucharitdas committed:
Updating the list to show as per new requirements : rave-less show one form (Capture Image) & rave show two forms (Form 1 and Form 2).

ssucharitdas committed:
-- Fix table rows number of lines, truncation on security question
-- Adding the validation text color on password and email view controller to red if any error.

ssucharitdas commented:
@marrodriguez @gthackway-mdsol 

## Validation Text in red:

<img width="352" alt="screen shot 2016-05-02 at 11 08 03 am" src="https://cloud.githubusercontent.com/assets/7715859/14958534/362e9f96-1059-11e6-8041-bc3a87c70b60.png">
<img width="352" alt="screen shot 2016-05-02 at 11 07 49 am" src="https://cloud.githubusercontent.com/assets/7715859/14958533/362dc832-1059-11e6-9953-ce55c791693f.png">
<img width="352" alt="screen shot 2016-05-02 at 11 07 27 am" src="https://cloud.githubusercontent.com/assets/7715859/14958532/362c470a-1059-11e6-86c4-fbb0b894b727.png">

## Table View text not truncated

<img width="352" alt="screen shot 2016-05-02 at 11 30 50 am" src="https://cloud.githubusercontent.com/assets/7715859/14958560/54040740-1059-11e6-957c-d688281b2bdb.png">
<img width="352" alt="screen shot 2016-05-02 at 11 30 44 am" src="https://cloud.githubusercontent.com/assets/7715859/14958561/54073f3c-1059-11e6-9518-5400aabdaa6b.png">

marrodriguez commented:
@ssucharitdas @gthackway-mdsol Password Confirmation, this text should be changed to "Your passwords do not match" (also in red)

ssucharitdas commented:
@marrodriguez I think email we can update to "Emails entered not matching" 
Not sure if we want to add words like "your", etc. Let me know your thoughts on this @yankeh

yankeh commented:
Let's use the same strings as what we have in the app now. That's the easiest way to go.

marrodriguez commented:
no, for email it is "Your emails do not match" (in red)

yankeh commented:
https://github.com/mdsol/Garuda/blob/develop/MedidataPRO/MedidataPRO/en.lproj/Localizable.strings

Bottom of the page.

marrodriguez commented:
There is a lot of white space between 'Select security question' and the answers, is this ok @gthackway-mdsol ?

marrodriguez commented:
👍  @yankeh thank you :)

ssucharitdas committed:
Updating email and password texts to match Garuda.

ssucharitdas committed:
Adding the offline mode error alert and also fixing the layout inset for table view [reducing gap between the header text and table].

ssucharitdas committed:
Adding the offline mode error alert and also fixing the layout inset for table view [reducing gap between the header text and table].

ncarroll-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -48,12 +57,37 @@ class CaptureImageViewController: UIViewController, UIImagePickerControllerDeleg
+                let client = clientFactory.clientOfType(MDClientType.Network);
+                var datastore = MDDatastoreFactory.create()
+                var subject = datastore.subjectWithID(self.subjectID)
+                subject.collectData(self.data, withMetadata: "Random String", withContentType: "", withAppSpecificTag: "", withSchemaURI: "", completion: { (dataEnvelope:  MDSubjectDataEnvelope!, err: NSError!) -> Void in
```
Can you add a content type? What's the actual data type being saved? image/png?

kbohlmann-mdsol commented on Podfile.lock:
```diff
@@ -20,30 +20,36 @@ PODS:
+  - AWSCore (2.3.6)
+  - AWSS3 (2.3.6):
+    - AWSCore (= 2.3.6)
+  - Babbage (2016.2.0.78):
```
I think this is still the wrong version of babbage?

ncarroll-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -77,14 +111,44 @@ class CaptureImageViewController: UIViewController, UIImagePickerControllerDeleg
+            let user = datastore.userWithID(self.userID)
+            
+            // Start an asynchronous task to load the subjects for the user logged in
+            client.loadSubjectsForUser(user) { (subjects: [AnyObject]!, error: NSError!) -> Void in
```
Why do we use [AnyObject]? Can this be changed to [MDSubject]?

ssucharitdas replied:
MDSubject throws error, so let us keep this as it is.

kbohlmann-mdsol replied:
The loadsubjects method in mdsubject.h just returns an nsarray, with no specifier, so I think this has to be AnyObject.

ncarroll-mdsol commented on AppConnectSwift/CaptureImageViewController.swift:
```diff
@@ -77,14 +111,44 @@ class CaptureImageViewController: UIViewController, UIImagePickerControllerDeleg
+                    }
+                }
+                else {
+                    NSOperationQueue.mainQueue().addOperationWithBlock {
```
Do these variables need to be set to nil on the main queue?

ssucharitdas replied:
Updated!

ncarroll-mdsol replied:
It looks like they're still nilled out in the main queue? Is this necessary?

ncarroll-mdsol commented on AppConnectSwift/FormListViewController.swift:
```diff
@@ -47,14 +47,19 @@ class FormListViewController: UITableViewController {
-                            self.spinner.stopAnimating()
-                            datastore = nil
-                            bgQueue = nil
+                    if subjects != nil {
```
When will subjects be nil?

ncarroll-mdsol commented on AppConnectSwift/PasswordViewController.swift:
```diff
@@ -0,0 +1,71 @@
+                }
+            }
+            catch {
+                print("Errored")
```
Can this be more specific?

ncarroll-mdsol replied:
Also, what error is the do...catch meant to catch?

ncarroll-mdsol commented on README.md:
```diff
@@ -30,6 +30,10 @@ Once the variables have been set, run `pod install` to install the necessary dep
 The application contains the following important view controllers
 
 - **LoginViewController** - Handles user authentication based on a provided username / password.
+- **EmailViewController** - First step of user registration to use AWS S3 storage, enter a valid email address and confirm the same.
```
I think the wording for these could be slightly different, maybe:
- EmailViewController: Handles user entry of the email address used to create a new account
- PasswordViewController: Handles user entry of the password used to create a new account
- SecurityQuestionViewController: Handles user selection of the security question used to create a new account
- CreateAccountViewController: Handles user entry of the security answer used to create a new account

?

ncarroll-mdsol commented on README.md:
```diff
@@ -9,9 +9,9 @@ AppConnectSwift is an example iOS app, written in Swift, that showcases proper u
 
 - You were provided Artifactory credentials by a Medidata representative.
-- You have a valid Rave installation with Patient Cloud functionality enabled.
+- You have a valid Rave installation with Patient Cloud functionality enabled.[For pushing data to Rave]
```
Maybe "(if you are using Rave-enabled forms)"?

ncarroll-mdsol commented on README.md:
```diff
@@ -45,7 +49,7 @@ To use this sample form:
 2. Log in with the sample app using the credentials of the subject you chose.
 
-> You should see two forms, Form 1 and Form 2. Form 1 opens as one page. Form 2 opens as multiple pages.
+> As general user sees an Image Capture form that opens a form for user to take a picture or load an image from library. As Rave registered user, addition to ImageCapture form you should see two more forms, Form 1 and Form 2. Form 1 opens as one page. Form 2 opens as multiple pages.
```
I think this probably belongs in a separate section, so we'd have:
- Using the Case Report Form (CRF)

and then
- Using the self-registration and data capture functionality

and the latter would have steps to register a user and upload data.

ncarroll-mdsol commented on README.md:
````diff
@@ -70,6 +74,14 @@ let key = "12345678901234567890123456789012".dataUsingEncoding(NSUTF8StringEncod
+
+```swift
+// In CreateAccountViewController.swift
+client.registerSubjectWithEmail(userEmail, password: userPassword, securityQuestionID: userSecurityQuestionID, securityQuestionAnswer: userSecurityAnswer completion:(void (^)(NSError *error))completion
````
Can this code be updated to use actual values, similar to how the other code examples are?

ssucharitdas committed:
Updating as per comments

ssucharitdas committed:
Updating the docs as per comment suggestion

ncarroll-mdsol commented on README.md:
```diff
@@ -67,15 +77,38 @@ In Swift / Objective-C, Babbage must be initialized with two arguments. The firs
+•  At least one upper-case letter
+•  At least one lower-case letter
+•  At least one numeric digit"
+let userSecurityQuestionID = "id of the array element from list of questions"
```
I think it would be best to use the correct data type (and above, a correctly formatted password) and then talk about the requirements in comments.

e.g. 

```
let userSecurityQuestionID = 1 // ID of the security question
```

ncarroll-mdsol commented on README.md:
````diff
@@ -67,15 +77,38 @@ In Swift / Objective-C, Babbage must be initialized with two arguments. The firs
 ```
 
-## Loading Data from the Datastore
+## Loading Data to the Datastore
````
I don't think this title could be a bit confusing. It looks like we have a TODO in the `Upload Data to S3` section, maybe it would be better under that header?

ncarroll-mdsol commented on README.md:
```diff
@@ -67,15 +77,38 @@ In Swift / Objective-C, Babbage must be initialized with two arguments. The firs
+subject.collectData(imageData, withMetadata: "Random String", withContentType: "image/png", withAppSpecificTag: "", withSchemaURI: "", completion: { (dataEnvelope:  MDSubjectDataEnvelope!, err: NSError!)}
+
+// Sending the dataEnvelop collected
+client.sendEnvelope(dataEnvelope, completion: {(err)}
```
Is this valid syntax, or does the handler need the type (NSError)?

ncarroll-mdsol commented on README.md:
```diff
@@ -67,15 +77,38 @@ In Swift / Objective-C, Babbage must be initialized with two arguments. The firs
+let userSecurityQuestionID = "id of the array element from list of questions"
+let userSecurityAnswer = "answer to the security question"
+
+client.registerSubjectWithEmail(userEmail, password: userPassword, securityQuestionID: userSecurityQuestionID, securityQuestionAnswer: userSecurityAnswer completion:(void (^)(NSError *error))completion
```
This looks like a mix of Objective-C and Swift syntax.

ncarroll-mdsol commented on README.md:
```diff
@@ -44,9 +48,15 @@ To use this sample form:
+### Using self-registration and data capture functionality
+
+1. User with valid api token, you can register a user.
+2. After registration user can upload different type of files.
```
These should be the steps to go through the process, just like the Rave section.
1. Use the in-app registration to enroll a user
2. Log in with the user that was just created
3. Select an image, which will be uploaded to AWS S3

ncarroll-mdsol commented on README.md:
```diff
@@ -67,15 +77,38 @@ In Swift / Objective-C, Babbage must be initialized with two arguments. The firs
+let imageData = UIImageJPEGRepresentation(img, 0.5)
+
+// Collecting the data from the image view
+subject.collectData(imageData, withMetadata: "Random String", withContentType: "image/png", withAppSpecificTag: "", withSchemaURI: "", completion: { (dataEnvelope:  MDSubjectDataEnvelope!, err: NSError!)}
```
It looks like the data is JPEG but you're sending `image/png` as the format - we probably want to make that consistent.

ssucharitdas committed:
Making the changes to make the code examples little more illustrative.

ssucharitdas committed:
Pushing to latest image context to work in compression.

ssucharitdas committed:
Add the datastore variable

ssucharitdas committed:
Removing the nil-ed out in else.

ssucharitdas committed:
Bumping babbage to 81 to cover end to end working

kbohlmann-mdsol commented:
ok, going in. 🍴
